Guidelines entities

attiks opened the pull request:
![Screenshot from 2021-11-18 19-32-43](https://user-images.githubusercontent.com/105144/142475746-7567190a-be70-4676-9da6-5d55f0aa2533.png)

New content entity created to not pollute the nodes, rough first version, but basically it's a field-able description for fields.

orakili commented:
@attiks Just to a note that guidelines for fields are just a tiny subset of the guidelines so I think having a "guideline" node type would be a better and easier way to go. 

This would notably allow to use the moderation backend for the editors to administer the guidelines, display them as single pages, use a view mode for the guidelines displayed in popup etc.

To link a guideline node to a field, we could instead create a field type with a widget that lists the available fields on content entity types and add such a field to the guideline node:

<img width="280" alt="Screen Shot 2021-11-19 at 17 28 24" src="https://user-images.githubusercontent.com/696348/142590700-cc9f6fc7-844d-45cc-aa67-ffe5f0f755de.png">

attiks commented:
![Screenshot from 2021-11-22 16-14-30](https://user-images.githubusercontent.com/105144/142886449-c6845622-6f68-4b0b-a1d4-1990e332f32f.png)

attiks commented:
Module split into generic and reliefweb specific, but [struggling with templates](https://github.com/UN-OCHA/rwint9-site/pull/186/files?authenticity_token=MU%2BITvQzdm8qhiICsClcQaylVPmBofQ1dU8SQgvlDyCKkFG2uhnVNDMX06D8VFgpj6sVTBzVRLI7g6ZoZb9O9Q%3D%3D&file-filters%5B%5D=.twig#diff-0a56ac36e1fd95d8b7e84235ae54c10eea8f9c1596847c43d3aa4a9d3cdf8377R14)

I assume there's a cleaner way

![Screenshot from 2021-11-24 16-54-58](https://user-images.githubusercontent.com/105144/143272021-d9dc97d1-4822-4020-b8be-e9c02d4925c7.png)

orakili commented:
@attiks The guidelines are only for internal usage so the page doesn't need to have the common header and footer (at least for now) so I think adding a `page--guidelines.html.twig` (or something like that based on the route) in the subtheme, that just displays the page content could do the trick.

attiks commented:
> page--guidelines.html.twig

Done, search added as well, but question was more about the use of `{{ guideline['#id'] }}` in the twig file

left23 commented:
I've added a few CSS changes to a separate branch, after merging the `develop` branch into this one.
https://github.com/UN-OCHA/rwint9-site/pull/191

left23 commented:
I've adjusted the frontend PR here https://github.com/UN-OCHA/rwint9-site/pull/194

left23 commented:
@attiks I'm adding a few screenshots below of things I noticed with working on this.

The section title or breadcrumb is absent (see D7 on right)
![Selection_171](https://user-images.githubusercontent.com/1835923/144884229-1fdc164d-5cf6-4162-8bfa-4b7544c19e85.png)

In some cases, links are not turned into anchor elements. This is inconsistent, some seem to work as intended.
![Selection_172](https://user-images.githubusercontent.com/1835923/144884295-cc754535-ccd3-44d5-8b8f-88b2f64dca62.png)

Some of the links that should be converted to the Guidelines are linking to the Trello card.
![Selection_174](https://user-images.githubusercontent.com/1835923/144884426-2232d072-52f4-44df-ab3e-ed24d911fe13.png)

Now that the title of the Guideline is showing (frontend PR #194 fixes the title display) it appears as `undefined`. The link works.
![Uploading Selection_176.png…]()

orakili commented on html/modules/custom/guidelines/src/Plugin/Field/FieldWidget/GuidelineFieldTargetWidget.php:
```diff
@@ -0,0 +1,139 @@
+
+            /** @var \Drupal\Core\Field\BaseFieldDefinition $field_definition */
+            $key = 'node.' . $bundle . '.' . $field_name;
+            $options[$key] = $bundle . ' > ' . $field_definition->getLabel() . ' (' . $field_name . ')';
```
It might be more user friendly to use the bundle label:

```php
$entity_type_manager = \Drupal::entityTypeManager();
$entity_type = $entity_type_manager
  ->getStorage($entity_type_id)
  ->getEntityType();
$bundle_label = $entity_type_manager
  ->getStorage($entity_type->getBundleEntityType())
  ->load($bundle)
  ->label();
```

orakili commented on html/modules/custom/guidelines/src/Plugin/Field/FieldWidget/GuidelineFieldTargetWidget.php:
```diff
@@ -0,0 +1,139 @@
+        $field_definitions = $entity_field_manager->getFieldDefinitions($entity_type, $bundle);
+        foreach ($field_definitions as $field_name => $field_definition) {
+          if ($field_definition instanceof BaseFieldDefinition) {
+            if (in_array($field_name, $hidden_base_fields)) {
```
In addition, maybe all the "computed", "internal" and "read-only" fields could be skipped.

orakili commented on html/modules/custom/reliefweb_entities/src/Services/JobFormAlter.php:
```diff
@@ -17,10 +17,6 @@ class JobFormAlter extends EntityFormAlterServiceBase {
    * {@inheritdoc}
    */
   protected function addBundleFormAlterations(array &$form, FormStateInterface $form_state) {
-    // Add a guide on how to populate the field as description to the title
-    // field.
-    $form['title']['widget'][0]['value']['#description'] = $this->t('The best job titles are brief and specific. Please refrain from indicating location, salary and other details in the title, if possible.');
-
```
Why remove this?

orakili commented on html/modules/custom/reliefweb_entities/src/Services/TrainingFormAlter.php:
```diff
@@ -19,10 +19,6 @@ class TrainingFormAlter extends EntityFormAlterServiceBase {
    * {@inheritdoc}
    */
   protected function addBundleFormAlterations(array &$form, FormStateInterface $form_state) {
-    // Add a guide on how to populate the field as description to the title
-    // field.
-    $form['title']['widget'][0]['value']['#description'] = $this->t('Should contain only the title of the training. Other information such as location, date and Organization should not be included in this field.');
-
```
Why remove this?

orakili commented on html/modules/custom/reliefweb_guidelines/src/Controller/GuidelineSinglePageController.php:
```diff
@@ -0,0 +1,429 @@
+        return $b['id'] <=> $a['id'];
+      });
+
+      $bundle = $query['bundle'];
+      $view = $query['view'] ?? '';
+      $exclude = $query['exclude'] ?? [];
+
+      $entities = RiverServiceBase::getRiverData($bundle, $result, $view, $exclude);
+      if (!empty($entities)) {
+        $build = [
+          '#theme' => 'reliefweb_rivers_river',
+          '#id' => '#headlines',
+          '#title' => $query['title'],
+          '#resource' => $query['resource'],
+          '#entities' => $entities,
+        ];
+      }
+    }
+
+    return new Response($this->renderer->render($build));
+  }
+
+  /**
+   * Ajax callback to update the list of selected headlines.
+   *
+   * @return \Symfony\Component\HttpFoundation\Response
+   *   200 if the update was successful, 500 otherwise.
+   */
+  public function updateHeadlines() {
+    $success = FALSE;
+    // Limit to 10,000 bytes (should never be reached).
+    $data = json_decode(file_get_contents('php://input', FALSE, NULL, 0, 10000) ?? '', TRUE);
+    if (!empty($data) && is_array($data)) {
+      $selection = array_slice(array_filter($data, 'is_int'), 0, 8);
+      $this->state->set('reliefweb_headline_selection', $selection);
+      $success = TRUE;
+    }
+    return new Response('', $success ? 200 : 500);
+  }
+
```
```suggestion
```

Remove the leftovers from copy/paste.

orakili replied:
```suggestion
```

Remove the leftovers from copy/paste.

orakili commented on html/modules/custom/reliefweb_guidelines/src/Controller/GuidelineSinglePageController.php:
```diff
@@ -0,0 +1,429 @@
+    return $build;
+  }
+
+  /**
```
Looks like all the methods below are leftovers from copying/pasting and can be removed.

orakili reviewed (commented):
Preliminary code review.

attiks commented on html/modules/custom/guidelines/guidelines.module:
```diff
@@ -0,0 +1,131 @@
+/**
+ * Implements hook_form_HOOK_alter().
+ */
+function guidelines_form_node_form_alter(array &$form, FormStateInterface $form_state) {
```
Added

attiks commented:
> The section title or breadcrumb is absent (see D7 on right)

Added

attiks commented:
> Some of the links that should be converted to the Guidelines are linking to the Trello card.

Fixed

attiks commented:
> In some cases, links are not turned into anchor elements. This is inconsistent, some seem to work as intended.

Fixed

left23 commented:
@orakili @attiks Note I've merged the frontend PR #194 into this branch now.

orakili commented:
@attiks Just a few more issues with the frontend part:

**Search**

Search on the /guidelines page is not working because the "search" input has the `search` id while the script expects that id to be on an empty article that will be used to show the search results:
<img width="370" alt="Screen Shot 2021-12-09 at 17 40 42" src="https://user-images.githubusercontent.com/696348/145362819-93e8d2d8-0077-4402-ac5d-753825525abd.png">
<img width="1060" alt="Screen Shot 2021-12-09 at 17 42 15" src="https://user-images.githubusercontent.com/696348/145362886-0055c6c4-b521-4a44-81e2-eaf7da8c3680.png">

**Sidebar**

The title in the sidebar should not have the breadcrumb:
<img width="343" alt="Screen Shot 2021-12-08 at 17 23 46" src="https://user-images.githubusercontent.com/696348/145363133-c4bd6531-1e98-4126-95ec-49a0338a0b8a.png">

**Popup**

The guidelines popups on the forms don't have a title:
<img width="431" alt="Screen Shot 2021-12-08 at 17 55 26" src="https://user-images.githubusercontent.com/696348/145366199-412409d8-dd4f-4cdd-8c7a-113d3cd6519b.png">

The link on the title should point to `/guidelines#xxxx`.

**Malformed HTML**

In the `/guidelines` page,  the text of all the guidelines after "Reports - Specifics > Submit Mailbox" is bolded:
<img width="1264" alt="Screen Shot 2021-12-09 at 18 09 04" src="https://user-images.githubusercontent.com/696348/145367028-b1d63141-e6e1-4a56-9bc5-d326aa4d6ef3.png">

**Lowercase fragment**

The fragments are all lower case (ex: `i8thikz4`) but the trello card ID and the existing fragment on the guidelines site are not. They should not be transformed for compatibility so that it's easier to compare the new guidelines with the exiting guidelines site and the trello cards.

**Image migration**

Some images seem to not be migrated properly (`0QdvgPyP`):
<img width="1169" alt="Screen Shot 2021-12-09 at 18 15 33" src="https://user-images.githubusercontent.com/696348/145368032-565f9f13-6696-4c34-8a97-d03ab6864791.png">

They exists on the guidelines site and in trello though.

attiks commented:
> Search on the /guidelines page is not working

Fixed

> The title in the sidebar should not have the breadcrumb

Fixed

> popups

Fixed

> fragments

Fixed

attiks commented:
> They exists on the guidelines site and in trello though.

And are fetched `html/sites/default/files/guidelines/trello/f4/03/f403ab69-e1ac-4182-88ba-10651c3c4211.png` but it isn't an image but an HTML file, source is https://monosnap.com/file/ntHKFOGBXob0k57KX2y6HTPyExoYa8

List of missing images

- https://d2r1vs3d9006ap.cloudfront.net/s3_images/1074644/Capture_inline.PNG?1404934476
- https://d2r1vs3d9006ap.cloudfront.net/s3_images/1074645/Capture_inline.PNG?1404934558
- https://trello-attachments.s3.amazonaws.com/554c519c9b98caa2495ea5db/55886da7f0e9839747835293/ea76364c871fe48ceeae43c29558462b/Desk.com%2BAgent%2B--%2BShuichi%2BOdaka%2B2015-08-05%2B15-42-04.png
- https://trello-attachments.s3.amazonaws.com/554c519c9b98caa2495ea5db/55886da7f0e9839747835293/27b7504f925ade60b031466ed7fbe884/Desk.com%2BAgent%2B--%2BShuichi%2BOdaka%2B2015-08-05%2B15-37-36.png
- https://trello-attachments.s3.amazonaws.com/55886da7f0e9839747835293/450x136/a8094d156323ad3d347634bf85477218/image11.png
- https://trello-attachments.s3.amazonaws.com/55886da7f0e9839747835293/536x174/0823345df190193f665673b6406db562/image00.png
- https://d2eos31sn1as5.cloudfront.net/30_01_12_16_32_30.png
- https://d2eos31sn1as5.cloudfront.net/30_01_12_16_37_28.png
- https://trello-attachments.s3.amazonaws.com/554c519c9b98caa2495ea5db/555ea52066e921d9d0953a80/4829683f12008a9c23a7378eeff16fbf/Notifications.PNG
- https://trello-attachments.s3.amazonaws.com/554c519c9b98caa2495ea5db/555ea52066e921d9d0953a80/3794f3cd420862cadae7a33bd0249135/leave_notifcations.PNG
- https://trello-attachments.s3.amazonaws.com/554c519c9b98caa2495ea5db/555ea52066e921d9d0953a80/2ab2f64a5be8eb9e609d4e2a13ee1f46/Training_notification.PNG
- https://trello.com/1/cards/613f588ea7c1cf0e0c1e9727/attachments/613f5946aba81d76e803dc63/download/Create_a_new_presentation_with_the_ReliefWeb_template_-_August_2021.gif
- https://trello.com/1/cards/613f588ea7c1cf0e0c1e9727/attachments/613f5944d16922359d4ca559/download/Create_a_new_document_with_the_ReliefWeb_template_-_August_2021.gif

attiks commented:
> Malformed HTML

Fixed

orakili commented on html/modules/custom/reliefweb_guidelines/src/Controller/GuidelineSinglePageController.php:
```diff
@@ -0,0 +1,171 @@
+  /**
+   * Clean A-tags.
+   */
+  protected function cleanAtags($text) {
```
The regex are too greedy (the `.+`) and this breaks some guidelines like the "Report - Posting ste-by-step > Headline" and following guideline. Also this convert links to trello cards from other boards.

Created https://humanitarian.atlassian.net/browse/RW-276 as follow-up ticket to fix those issues.

orakili reviewed (approved):
Looks good, thanks. 

There are still 2 main: the link conversion which is still too greedy and breaks some pages (see RW-276) and some field mappings are missing (see RW-277).

The PR is already really large so going to merge this and let's handle the follow-ups in other PRs.
